File: src/spor/diff.py
import difflib

from .anchor import make_anchor


def _split_keep_sep(s, sep):
    toks = s.split(sep)
    result = [tok + sep for tok in toks[:-1]]
    result.append(toks[-1])
    return result


def _context_diff(file_name, c1, c2):
    c1_text = _split_keep_sep(c1.full_text, '\n')
    c2_text = _split_keep_sep(c2.full_text, '\n')

    return difflib.context_diff(
        c1_text, c2_text,
        fromfile='{} [original]'.format(file_name),
        tofile='{} [current]'.format(file_name))


def get_anchor_diff(anchor):
    """Get the get_anchor_diff between an anchor and the current state of its source.

    Returns: A tuple of get_anchor_diff lines. If there is not different, then this
        returns an empty tuple.
    """
    new_anchor = make_anchor(
        file_path=anchor.file_path,
        offset=anchor.context.offset,
        width=len(anchor.context.topic),
        context_width=anchor.context.width,
        metadata=anchor.metadata)

    assert anchor.file_path == new_anchor.file_path
    assert anchor.context.offset == new_anchor.context.offset
    assert len(anchor.context.topic) == len(new_anchor.context.topic)
    assert anchor.metadata == new_anchor.metadata

    return tuple(
        _context_diff(
            anchor.file_path,
            anchor.context,
            new_anchor.context))

File: src/diff.rs
use crate::anchor;
use crate::file_io::read_file;
use thiserror::Error;

/// Check to see if and how an anchored file has changed since an anchor was created.
///
/// Get the difference between an existing anchor and an anchor built against the current state
/// of its file.
///
/// # Arguments
///
/// * `anchor` - The existing anchor to check.
///
/// # Returns
///
/// The first element of the returned tuple indicates whether any changes were found. The
/// second element is a vector of strings describing the diff (this may have contents even
/// if there is no actual difference).
/// 
/// # Errors
///
/// A std::error::Error describing any failure to construct a new Anchor.
pub fn get_anchor_diff(anchor: &anchor::Anchor) -> Result<(bool, Vec<String>), DiffError> {
    let full_text = read_file(anchor.file_path(), anchor.encoding())?;

    let context = anchor::Context::from_text(
        &full_text,
        anchor.context().offset(),
        anchor.context().topic().len(),
        anchor.context().width(),
    ).map_err(|err| DiffError::InvalidAnchor(err.to_string()))?;

    let new_anchor = anchor::Anchor::new(
        anchor.file_path(),
        context,
        anchor.metadata().clone(),
        anchor.encoding().clone(),
    ).map_err(|err| DiffError::InvalidAnchor(err.to_string()))?;

    let mut diff_strings: Vec<String> = Vec::new();

    let mut changed = false;

    for diff in diff::lines(
        anchor.context().full_text().as_str(),
        new_anchor.context().full_text().as_str(),
    ) {
        changed = match diff {
            diff::Result::Both(_, _) => changed,
            _ => true,
        };

        let diff_text = match diff {
            diff::Result::Left(l) => format!("-{}", l),
            diff::Result::Both(l, _) => format!(" {}", l),
            diff::Result::Right(r) => format!("+{}", r),
        };
        diff_strings.push(diff_text);
    }

    Ok((changed, diff_strings))
}

#[derive(Error, Debug)]
pub enum DiffError {
    #[error("Unable to create new anchor: {0}")]
    InvalidAnchor(String),

    #[error(transparent)]
    Io {
        #[from]
        source: std::io::Error,
    },

}

